replications/remotewrite/writer.go

```diff
@@ -0,0 +1,242 @@
+		waitFunc: func(t time.Duration) <-chan time.Time {
+			return time.After(t)
+		},
+	}
+}
+
+func (w *writer) Write(ctx context.Context, data []byte) error {
+	attempts := 0
+
+	for {
+		if w.maximumAttemptsBeforeErr > 0 && attempts >= w.maximumAttemptsBeforeErr {
+			return errors.New("maximum number of attempts exceeded")
+		}
+
+		// Get the most recent config on every attempt, in case the user has updated the config to correct errors.
+		conf, err := w.configStore.GetFullHTTPConfig(ctx, w.replicationID)
+		if err != nil {
+			return err
+		}
+
+		res, err := PostWrite(ctx, conf, data, w.clientTimeout)
+		res, msg, ok := normalizeResponse(res, err)
+		if !ok {
+			// Can't retry - bail out.
+			return err
+		}
+
+		if err := w.configStore.UpdateResponseInfo(ctx, w.replicationID, res.StatusCode, msg); err != nil {
+			return err
+		}
+
+		if err == nil {
+			// Successful write
+			return nil
+		}
+
+		attempts++
+		var waitTime time.Duration
+		hasSetWaitTime := false
+
+		switch res.StatusCode {
+		case http.StatusBadRequest:
+			if conf.DropNonRetryableData {
+				w.logger.Debug(fmt.Sprintf("dropped %d bytes of data due to %d response from server", len(data), http.StatusBadRequest))
+				return nil
+			}
+		case http.StatusTooManyRequests:
+			headerTime := w.waitTimeFromHeader(res)
+			if headerTime != 0 {
+				waitTime = headerTime
+				hasSetWaitTime = true
+			}
+		}
+
+		if !hasSetWaitTime {
+			waitTime = w.backoff(attempts)
+		}
+
+		select {
+		case <-w.waitFunc(waitTime):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// normalizeReponse returns a guaranteed non-nil value for *http.Response, and an extracted error message string for use
+// in logging. The returned bool indicates that the response is retryable - false means that the write request should be
+// aborted due to a malformed request.
+func normalizeResponse(r *http.Response, err error) (*http.Response, string, bool) {
+	var errMsg string
+	if err != nil {
+		errMsg = err.Error()
+	}
+
+	if r == nil {
+		if errorIsTimeout(err) {
+			return &http.Response{}, errMsg, true
+		}
+
+		return &http.Response{}, errMsg, false
+	}
+
+	return r, errMsg, true
+}
+
+func errorIsTimeout(err error) bool {
+	if err, ok := err.(net.Error); ok && err.Timeout() {
+		return true
+	}
+
+	return false
+}
+
+func PostWrite(ctx context.Context, config *influxdb.ReplicationHTTPConfig, data []byte, timeout time.Duration) (*http.Response, error) {
+	u, err := url.Parse(config.RemoteURL)
+	if err != nil {
+		return nil, invalidRemoteUrl(config.RemoteURL, err)
+	}
+
+	params := api.ConfigParams{
+		Host:             u,
+		UserAgent:        userAgent,
+		Token:            &config.RemoteToken,
+		AllowInsecureTLS: config.AllowInsecureTLS,
+	}
+	conf := api.NewAPIConfig(params)
+	conf.HTTPClient.Timeout = timeout
+	client := api.NewAPIClient(conf).WriteApi
+
+	req := client.PostWrite(ctx).
+		Org(config.RemoteOrgID.String()).
+		Bucket(config.RemoteBucketID.String()).
```

Comment:
Is this the "normal" way to format chained method calls? It could very well be, it just looks strange to me being used to something like:
```
client.PostWrite(ctx)
        .Org(config.RemoteOrgID.String())
        .Bucket(config.RemoteBucketID.String())
        ...
```

Reply:
I agree that it looks weird! But it doesn't compile with the dots on the newlines. So as far as I know that's the only way for multi-line chained method calls to work.